File: src/backend/Intelibill.Application/Features/Sales/DTOs/SellableDto.cs
namespace Intelibill.Application.Features.Sales.DTOs;

public enum SellableKind
{
    Goods,
    Service,
}

public sealed record SellableDto(
    SellableKind Kind,
    Guid? InventoryBatchId = null,
    string? Barcode = null,
    string? ItemName = null,
    string? BatchNumber = null,
    decimal? Quantity = null,
    decimal? SalesPrice = null,
    decimal? Mrp = null,
    decimal? TaxRatePercent = null,
    bool? TaxIncluded = null,
    bool? PurchaseTaxIncluded = null,
    DateOnly? ExpiryDate = null,
    Guid? ServiceId = null,
    string? Code = null,
    string? Name = null,
    string? Description = null,
    decimal? Price = null,
    string? HsnCode = null)
{
    public static SellableDto FromInventoryBatch(
        Guid inventoryBatchId,
        string barcode,
        string itemName,
        string batchNumber,
        decimal quantity,
        decimal salesPrice,
        decimal mrp,
        decimal taxRatePercent,
        bool taxIncluded,
        bool purchaseTaxIncluded,
        DateOnly? expiryDate) =>
        new(
            SellableKind.Goods,
            InventoryBatchId: inventoryBatchId,
            Barcode: barcode,
            ItemName: itemName,
            BatchNumber: batchNumber,
            Quantity: quantity,
            SalesPrice: salesPrice,
            Mrp: mrp,
            TaxRatePercent: taxRatePercent,
            TaxIncluded: taxIncluded,
            PurchaseTaxIncluded: purchaseTaxIncluded,
            ExpiryDate: expiryDate);

    public static SellableDto FromService(
        Guid serviceId,
        string code,
        string name,
        string? description,
        decimal price,
        string? hsnCode,
        decimal taxRatePercent,
        bool taxIncluded) =>
        new(
            SellableKind.Service,
            ServiceId: serviceId,
            Code: code,
            Name: name,
            Description: description,
            Price: price,
            HsnCode: hsnCode,
            TaxRatePercent: taxRatePercent,
            TaxIncluded: taxIncluded);
}

File: src/mobile/android/intelibill_mobile/lib/src/features/sales/data/dto/sellable_dto.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'sellable_dto.freezed.dart';
part 'sellable_dto.g.dart';

@freezed
sealed class SellableDto with _$SellableDto {
  const factory SellableDto({
    @JsonKey(name: 'kind') required String kind,
    @JsonKey(name: 'inventoryBatchId') String? inventoryBatchId,
    @JsonKey(name: 'barcode') String? barcode,
    @JsonKey(name: 'itemName') String? itemName,
    @JsonKey(name: 'batchNumber') String? batchNumber,
    @JsonKey(name: 'quantity') @Default(0.0) double quantity,
    @JsonKey(name: 'salesPrice') @Default(0.0) double salesPrice,
    @JsonKey(name: 'mrp') @Default(0.0) double mrp,
    @JsonKey(name: 'taxRatePercent') @Default(0.0) double taxRatePercent,
    @JsonKey(name: 'taxIncluded') @Default(false) bool taxIncluded,
    @JsonKey(name: 'purchaseTaxIncluded')
    @Default(false)
    bool purchaseTaxIncluded,
    @JsonKey(name: 'expiryDate') DateTime? expiryDate,
    @JsonKey(name: 'serviceId') String? serviceId,
    @JsonKey(name: 'code') String? code,
    @JsonKey(name: 'name') String? name,
    @JsonKey(name: 'description') String? description,
    @JsonKey(name: 'price') @Default(0.0) double price,
    @JsonKey(name: 'hsnCode') String? hsnCode,
  }) = _SellableDto;

  factory SellableDto.fromJson(Map<String, dynamic> json) =>
      _$SellableDtoFromJson(json);
}
